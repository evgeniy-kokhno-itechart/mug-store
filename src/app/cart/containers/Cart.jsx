import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import CartTable from '../components/CartTable';
import TotalCost from '../components/TotalCost';
import { cartCostsSelector } from '../CartSelectors';
import { cartActions } from '../CartActions';

export class Cart extends Component {
  renderCartDetails(products) {
    const {
      currentUserName, cart, isCurrencyLoading, changeQuantity, deleteProductFromCart,
    } = this.props;
    return (
      <React.Fragment>
        <CartTable
          sortColumn="id"
          productsInCart={products}
          isCurrencyLoading={isCurrencyLoading}
          changeQuantity={changeQuantity}
          deleteProductFromCart={deleteProductFromCart}
        />
        <div className="row justify-content-between mx-2">
          <TotalCost total={cart.totalCost} customClasses="h5 my-auto" />
          <Link className="btn btn-secondary justify-content-end" to={currentUserName ? '/order/mycart' : '/order'}>
            Order Now!
          </Link>
        </div>
      </React.Fragment>
    );
  }

  renderEmptyMessage() {
    return <h1 className="m-3 text-center">Your cart is empty</h1>;
  }

  render() {
    const { cart } = this.props;

    return cart.products.length
      ? this.renderCartDetails(cart.products)
      : this.renderEmptyMessage();
  }
}

Cart.propTypes = {
  cart: PropTypes.shape({
    products: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        imageURL: PropTypes.string,
        title: PropTypes.string,
        description: PropTypes.string,
        currentCurrencyCost: PropTypes.number,
        quantity: PropTypes.number,
      }),
    ),
    totalCost: PropTypes.number,
  }).isRequired,

  isCurrencyLoading: PropTypes.bool,

  currentUserName: PropTypes.string,

  changeQuantity: PropTypes.func.isRequired,
  deleteProductFromCart: PropTypes.func.isRequired,
};

Cart.defaultProps = {
  isCurrencyLoading: true,
  currentUserName: '',
};

const mapStateToProps = state => ({
  cart: cartCostsSelector(state),
  currentUserName: state.user.currentUser.name,
  isCurrencyLoading: state.currency.currenciesStatus.isInProcess,
});

const mapDispatchToProps = {
  changeQuantity: cartActions.ChangeQuantity,
  deleteProductFromCart: cartActions.DeleteFromCart,
};

export default connect(mapStateToProps, mapDispatchToProps)(Cart);
