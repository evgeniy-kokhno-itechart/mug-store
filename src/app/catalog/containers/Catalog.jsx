/* eslint-disable no-nested-ternary */
import React, { Component } from 'react';
import _ from 'lodash';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { categoryActions } from '../CategoryActions';
import paginate from '../Paginate';
import CatalogTable from '../components/CatalogTable';
import CatalogTableHeader from '../components/CatalogTableHeader';
import CatalogTableFooter from '../components/CatalogTableFooter';
import { ListGroup, ErrorMessage, Spinner } from '../../shared';
import { // deleteProduct, getProducts,
  productsPricesSelector,
  productsActions,
} from '../../product';
import { cartActions } from '../../cart';

export class Catalog extends Component {
  state = {
    searchQuery: '',
    sortColumn: { id: 'title_asc', path: 'title', order: 'asc' },
    pageSize: 10,
    currentPage: 1,
  };

  pageSizeOptions = [{ id: '5', name: '5' }, { id: '10', name: '10' }, { id: '15', name: '15' }];

  sortOptions = [
    { id: 'currentCurrencyPrice_asc', name: 'Price A-Z' },
    { id: 'currentCurrencyPrice_desc', name: 'Price Z-A' },
    { id: 'title_asc', name: 'Title A-Z' },
    { id: 'title_desc', name: 'Title Z-A' },
    { id: 'rate_asc', name: 'Rate 1-5' },
    { id: 'rate_desc', name: 'Rate 5-1' },
  ];

  allProductsCategory = { id: '', name: 'All Products' };

  componentDidMount() {
    this.props.getProducts();
  }

  displayIsLoading = () => {
    const { products, category } = this.props;
    if (products.catalogProductsStatus.isInProcess || category.loadingStatus.isInProcess || products.productDeletingStatus.isInProcess) {
      return <Spinner spinnerClasses="spinner--big" wrapperClasses="mt-5" />;
    }
    return null;
  }

  displayGotError = () => {
    const { products, category } = this.props;
    if (products.catalogProductsStatus.hasFailed || category.loadingStatus.hasFailed || products.productDeletingStatus.hasFailed) {
      return (
        <ErrorMessage message={products.catalogProductsStatus.error || category.loadingStatus.error || category.loadingStatus.error} />
      );
    }
    return null;
  }

  displayPagedDataOrEmpty = () => {
    const { totalCount, productsOnPage } = this.getPagedData();
    return productsOnPage.length
      ? this.renderProducts(totalCount, productsOnPage)
      : this.renderEmptyMessage();
  }

  getPagedData = () => {
    const {
      pageSize, currentPage, sortColumn, searchQuery,
    } = this.state;

    const { category } = this.props;

    const allProducts = this.props.products.productsForCatalog;
    let filtered = allProducts;
    if (searchQuery) {
      filtered = allProducts.filter(m => m.title.toLowerCase().includes(searchQuery.toLowerCase()));
    } else if (category.currentCategory && category.currentCategory.id) {
      filtered = allProducts.filter(m => m.category.id === category.currentCategory.id);
    }

    const sorted = _.orderBy(filtered, [sortColumn.path], [sortColumn.order]);

    const productsOnPage = paginate(sorted, currentPage, pageSize);

    return { totalCount: filtered.length, productsOnPage };
  };

  handleItemSelect = (category) => {
    this.setState({ searchQuery: '', currentPage: 1 });
    this.props.changeCategory(category);
  };

  handlePageChange = (page) => {
    this.setState({ currentPage: page });
  };

  handleSearch = (e, query) => {
    e.preventDefault();
    this.props.changeCategory(this.allProductsCategory);
    this.setState({ searchQuery: query, currentPage: 1 });
  };

  handleSort = (e) => {
    const detailsToBeSplitted = e.currentTarget.value;
    if (detailsToBeSplitted) {
      const sortInfo = detailsToBeSplitted.split('_');

      this.setState({ sortColumn: { id: detailsToBeSplitted, path: sortInfo[0], order: sortInfo[1] } });
    } else {
      this.setState({ sortColumn: { path: 'title', order: 'asc' } });
    }
  };

  handleItemsCountChange = (e) => {
    const pageSizeString = e.currentTarget.value;
    const pageSize = +pageSizeString;
    this.setState({ pageSize, currentPage: 1 });
  };

  handleDelete = (productId) => {
    this.props.deleteProduct(productId);
    this.props.getProducts();
  };

  renderEmptyMessage = () => (
    <h1 className="m-2 text-center">
      There are no products in the catalog
      <div className="m-1 m-md-3 mx-auto">
        <Link className="btn btn-secondary" to="/">
          To Main Page
        </Link>
      </div>
    </h1>
  );

  renderProducts = (totalCount, productsOnPage) => {
    const {
      sortColumn, searchQuery, pageSize, currentPage,
    } = this.state;

    const {
      // eslint-disable-next-line no-shadow
      currentUserRoles, category, isCurrencyLoading, addToCart,
    } = this.props;

    return (
      <div className="row mt-3">
        <div className="col-11 col-sm-3 col-lg-2 mx-auto mb-2">
          <ListGroup
            items={[this.allProductsCategory, ...category.categories]}
            selectedItem={category.currentCategory}
            onItemSelect={this.handleItemSelect}
          />
        </div>

        <div className="col-12 col-sm-9 col-lg-10">
          <CatalogTableHeader
            currentUserRoles={currentUserRoles}
            searchQuery={searchQuery}
            sortColumnKey={sortColumn.id}
            sortOptions={this.sortOptions}
            handleSearch={this.handleSearch}
            handleSort={this.handleSort}
          />

          <CatalogTable
            sortColumn={sortColumn}
            productsOnPage={productsOnPage}
            currentUserRoles={currentUserRoles}
            isCurrencyLoading={isCurrencyLoading}
            addToCart={addToCart}
            onDelete={this.handleDelete}
          />

          <CatalogTableFooter
            totalCount={totalCount}
            pageSize={pageSize}
            pageSizeOptions={this.pageSizeOptions}
            currentPage={currentPage}
            onPageChange={this.handlePageChange}
            onItemsCountChange={this.handleItemsCountChange}
          />
        </div>
      </div>
    );
  };

  render() {
    return this.displayIsLoading() || this.displayGotError() || this.displayPagedDataOrEmpty();
  }
}

Catalog.propTypes = {
  currentUserRoles: PropTypes.arrayOf(PropTypes.string),

  products: PropTypes.shape({
    productsForCatalog: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        imageURL: PropTypes.string,
        title: PropTypes.string,
        description: PropTypes.string,
        category: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
        basePrice: PropTypes.number,
        currentCurrencyPrice: PropTypes.number,
        discount: PropTypes.number,
        producer: PropTypes.string,
        publishDate: PropTypes.string,
        rate: PropTypes.string,
      }),
    ),
    catalogProductsStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
    productDeletingStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
  }),

  category: PropTypes.shape({
    categories: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })),
    currentCategory: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
    loadingStatus: PropTypes.shape({ isInProcess: PropTypes.bool, hasFailed: PropTypes.bool, error: PropTypes.string }),
  }),

  isCurrencyLoading: PropTypes.bool,

  getProducts: PropTypes.func.isRequired,
  deleteProduct: PropTypes.func.isRequired,
  changeCategory: PropTypes.func.isRequired,
  addToCart: PropTypes.func.isRequired,
};

Catalog.defaultProps = {
  currentUserRoles: [],

  products: {
    productsForCatalog: [],
    productsGettingStatus: { isInProcess: false, hasFailed: false, error: '' },
    productDeletingStatus: { isInProcess: false, hasFailed: false, error: '' },
  },

  category: {
    categories: [],
    currentCategory: {},
    loadingStatus: { isInProcess: false, hasFailed: false, error: '' },
  },

  isCurrencyLoading: true,
};

const mapStateToProps = state => ({
  currentUserRoles: state.user.currentUser.roles,

  products: {
    productsForCatalog: productsPricesSelector(state),
    catalogProductsStatus: state.products.catalogProductsStatus,
    productDeletingStatus: state.products.deletingStatus,
  },

  isCurrencyLoading: state.currency.currenciesStatus.isInProcess,

  category: state.category,
});

const mapDispatchToProps = {
  changeCategory: categoryActions.ChangeCategory,
  getProducts: productsActions.GetProducts.InitiateApiCall,
  deleteProduct: productsActions.DeleteProduct.InitiateApiCall,
  addToCart: cartActions.AddToCart,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Catalog);
