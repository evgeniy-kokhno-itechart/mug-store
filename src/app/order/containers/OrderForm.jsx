import React, { Component } from 'react';
import * as Yup from 'yup';
import _ from 'lodash';
import { PropTypes } from 'prop-types';
import { Input, TextArea, FormService } from '../../shared';

class OrderForm extends Component {
  state = {
    data: {
      name: '',
      country: '',
      city: '',
      address: '',
      phone: '',
      comment: '',
    },
    errors: {},
  };

  orderFormSchema = {
    name: Yup.string()
      .required()
      .max(100)
      .label('Name'),
    country: Yup.string()
      .required()
      .max(100)
      .label('Country'),
    city: Yup.string()
      .required()
      .max(100)
      .label('City'),
    address: Yup.string()
      .required()
      .max(250)
      .label('Address'),
    phone: Yup.string()
      .required()
      .min(12)
      .max(17)
      .label('Phone'),
    comment: Yup.string().label('Comment'),
  };

  componentDidMount() {
    if (this.props.currentUser.name) {
      this.setState({
        data: this.mapToViewModel(this.props.currentUser),
      });
    }
  }

  handleChange = (e) => {
    const { target: input } = e;

    // null stands for matchedInputName
    this.setState(prevState => FormService.handleChange(input, null, prevState, this.orderFormSchema));
  };

  handleSubmit = (e) => {
    e.preventDefault();
    this.props.onOrderSubmit('/orderconfirm');
  };

  mapToViewModel({
    name, country, city, address, phone,
  }) {
    return {
      name,
      country,
      city,
      address,
      phone,
    };
  }

  render() {
    const { data, errors } = this.state;
    return (
      <React.Fragment>
        <p>Please fill out the form to continue</p>
        <form onSubmit={this.handleSubmit}>
          <Input type="text" name="name" label="Name" value={_.get(data, 'name')} error={errors.name} onValueChange={this.handleChange} />

          <Input
            type="text"
            name="country"
            label="Country"
            value={_.get(data, 'country')}
            error={errors.country}
            onValueChange={this.handleChange}
          />

          <Input type="text" name="city" label="City" value={_.get(data, 'city')} error={errors.city} onValueChange={this.handleChange} />

          <Input
            type="text"
            name="address"
            label="Address"
            value={_.get(data, 'address')}
            error={errors.address}
            onValueChange={this.handleChange}
          />

          <Input
            type="text"
            name="phone"
            label="Phone"
            value={_.get(data, 'phone')}
            error={errors.phone}
            onValueChange={this.handleChange}
          />

          <TextArea name="comment" label="Comment" value={data.comment} error={errors.comment} onValueChange={this.handleChange} />

          <button
            type="submit"
            disabled={FormService.validateForm(this.orderFormSchema, data)}
            className="btn btn-secondary d-block mx-auto mx-md-0"
          >
            Submit Order
          </button>
        </form>
      </React.Fragment>
    );
  }
}

OrderForm.propTypes = {
  currentUser: PropTypes.shape({
    id: PropTypes.string,
    address: PropTypes.string,
    city: PropTypes.string,
    country: PropTypes.string,
    name: PropTypes.string,
    phone: PropTypes.string,
    roles: PropTypes.arrayOf(PropTypes.string),
    username: PropTypes.string,
  }),
  onOrderSubmit: PropTypes.func.isRequired,
};

OrderForm.defaultProps = {
  currentUser: {},
};

export default OrderForm;
