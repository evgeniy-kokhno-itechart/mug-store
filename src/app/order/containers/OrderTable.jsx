import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import { Table } from '../../shared';
import { ProductPrice } from '../../cart';

export class OrderTable extends Component {
  columns = [
    {
      key: 'title',
      label: 'Title',
      content: product => (
        <Link to={`/products/${product.id}`} className="clickable">
          {product.title}
        </Link>
      ),
      customClasses: 'order_table__column--title',
    },
    {
      path: 'quantity',
      label: 'Quantity',
      customClasses: 'text-center',
    },
    {
      key: 'cost',
      label: 'Cost',
      content: product => <ProductPrice price={product.currentCurrencyCost} isCurrencyLoading={this.props.isCurrencyLoading} />,
      customClasses: 'text-center order_table__column--cost',
    },
  ];

  render() {
    const { products, sortColumn } = this.props;

    return (
      <React.Fragment>
        <p>Please check the list of products</p>
        <Table columns={this.columns} sortColumn={sortColumn} items={products} />
      </React.Fragment>
    );
  }
}

OrderTable.propTypes = {
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      quantity: PropTypes.number,
      price: PropTypes.object,
      title: PropTypes.string,
    }),
  ).isRequired,
  sortColumn: PropTypes.string.isRequired,

  isCurrencyLoading: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  isCurrencyLoading: state.currency.currenciesStatus.isGettingCurrenciesInProcess,
});

export default connect(mapStateToProps)(OrderTable);
