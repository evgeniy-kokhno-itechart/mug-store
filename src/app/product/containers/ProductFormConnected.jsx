import React from 'react';
import { PropTypes } from 'prop-types';
import { connect } from 'react-redux';
import { productsActions } from '../ProductsActions';
import { productCostSelector } from '../ProductsSelectors';
import { FormBase, Spinner, ErrorMessage } from '../../shared';
import ProductForm from '../components/ProductForm/ProductForm';
import { CurrencyService } from '../../currency';


export class ProductFormConnected extends FormBase {
  // define state and componentDidUpdate to avoid direct mapping props to child when productId is "new"
  // and componentWillUnmount wasn't called so clearCurrentProductInfo wasn't call as well
  state = { data: {} }

  componentDidMount() {
    const productId = this.props.match.params.id;
    if (productId === 'new') {
      return;
    }

    this.props.getProduct(productId);
  }

  componentDidUpdate(prevProps, prevState) {
    const { productState } = this.props;

    if (prevState.data.id !== productState.product.id) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ data: this.mapToViewModel(this.props.productState.product) });
    }
  }

  displayIsLoading = () => {
    const { productState } = this.props;
    if (productState.loadingStatus.isInProcess || productState.savingStatus.isInProcess) {
      return <Spinner spinnerClasses="spinner--medium" wrapperClasses="mt-5" />;
    }
    return null;
  }

  displayGotError = () => {
    const { productState } = this.props;
    if (productState.loadingStatus.hasFailed || productState.savingStatus.hasFailed) {
      return <ErrorMessage message={productState.loadingStatus.error || productState.savingStatus.error} />;
    }
    return null;
  }

  handleSubmit = (productViewModel) => {
    const product = this.mapFromViewModel(productViewModel);
    this.props.saveProduct({ product, redirectUrl: '/catalog' });
  };

  mapToViewModel(product) {
    return {
      id: product.id,
      title: product.title,
      imageURL: product.imageURL,
      description: product.description,
      categoryId: product.category.id,
      currentCurrencyPrice: product.currentCurrencyPrice,
      discount: product.discount,
      producer: product.producer,
      publishDate: product.publishDate,
      rate: product.rate,
    };
  }

  mapFromViewModel(data) {
    return {
      id: data.id,
      title: data.title,
      imageURL: data.imageURL,
      description: data.description,
      category: this.props.categories.find(category => category.id === data.categoryId),
      basePrice: CurrencyService.getBasePrice(data.currentCurrencyPrice, this.props.currentCurrency.rate),
      discount: data.discount,
      producer: data.producer,
      rate: data.rate,
    };
  }

  renderForm = () => {
    const { categories, currentCurrency } = this.props;
    return (
      <ProductForm
        currentCurrencyName={currentCurrency.name}
        categories={categories}
        product={this.state.data}
        onSubmit={this.handleSubmit}
      />
    );
  };

  render() {
    return this.displayIsLoading() || this.displayGotError() || this.renderForm();
  }
}

ProductFormConnected.propTypes = {
  match: PropTypes.shape({ params: PropTypes.shape({ id: PropTypes.string }) }),

  currentCurrency: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    rate: PropTypes.number,
  }).isRequired,

  categories: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })).isRequired,

  productState: PropTypes.shape({
    product: PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      imageURL: PropTypes.string.isRequired,
      category: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
      description: PropTypes.string,
      producer: PropTypes.string,
      discount: PropTypes.number,
      currentCurrencyPrice: PropTypes.number,
      rate: PropTypes.string,
    }),
    loadingStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
    savingStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
  }),

  getProduct: PropTypes.func.isRequired,
  saveProduct: PropTypes.func.isRequired,
};

ProductFormConnected.defaultProps = {
  categories: [],
  match: { params: { id: '' } },

  productState: {
    product: {
      id: '',
      imageURL: '',
      title: 'defaultProduct',
      description: '',
      category: {},
      currentCurrencyPrice: 0,
      discount: 0,
      producer: '',
      rate: '',
    },
    loadingStatus: {
      isInProcess: true,
      hasFailed: false,
      error: '',
    },
    savingStatus: {
      isInProcess: true,
      hasFailed: false,
      error: '',
    },
  },
};

const mapStateToProps = state => ({
  currentCurrency: state.currency.currentCurrency,
  categories: state.category.categories,

  productState: {
    product: productCostSelector(state),
    loadingStatus: state.products.currentProductStatus,
    savingStatus: state.products.savingStatus,
  },
});

const mapDispatchToProps = {
  getProduct: productsActions.GetProduct.InitiateApiCall,
  saveProduct: productsActions.SaveProduct.InitiateApiCall,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProductFormConnected);
