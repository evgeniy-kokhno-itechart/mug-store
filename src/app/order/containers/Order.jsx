import React, { Component } from 'react';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import OrderTable from '../components/OrderTable';
import OrderForm from '../components/OrderForm';
import { TotalCost, cartActions, cartCostsSelector } from '../../cart';

export class Order extends Component {
  submitOrder = () => {
    this.props.submitCartOrder('/orderconfirm');
  }

  render() {
    const { cart, currentUser, isCurrencyLoading } = this.props;
    return (
      <React.Fragment>
        <h1 className="text-center m-3">Please check your order</h1>
        <div className="row m-2">
          <div className="col-md-6">
            <OrderTable sortColumn="title" products={cart.products} isCurrencyLoading={isCurrencyLoading} />
            <TotalCost total={cart.totalCost} customClasses="float-right" />
          </div>
          <div className="col-md-5 offset-md-1">
            <OrderForm currentUser={currentUser} onSubmit={this.submitOrder} />
          </div>
        </div>
      </React.Fragment>
    );
  }
}

Order.propTypes = {
  cart: PropTypes.shape({
    products: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        quantity: PropTypes.number,
        price: PropTypes.object,
        title: PropTypes.string,
      }),
    ),
    totalCost: PropTypes.number,
  }).isRequired,
  currentUser: PropTypes.shape({
    id: PropTypes.string,
    address: PropTypes.string,
    city: PropTypes.string,
    country: PropTypes.string,
    name: PropTypes.string,
    phone: PropTypes.string,
    roles: PropTypes.arrayOf(PropTypes.string),
    username: PropTypes.string,
  }),

  isCurrencyLoading: PropTypes.bool,

  submitCartOrder: PropTypes.func.isRequired,
};

Order.defaultProps = {
  currentUser: {},
  isCurrencyLoading: true,
};

const mapStateToProps = state => ({
  cart: cartCostsSelector(state),
  currentUser: state.user.currentUser,
  isCurrencyLoading: state.currency.currenciesStatus.isInProcess,
});

const mapDipatchToProps = {
  submitCartOrder: cartActions.SubmitCartOrder.InitiateApiCall,
};

export default connect(
  mapStateToProps,
  mapDipatchToProps,
)(Order);
