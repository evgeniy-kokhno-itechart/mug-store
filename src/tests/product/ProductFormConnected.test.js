/* eslint-disable no-undef */
import React from 'react';

import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { ProductFormConnected } from '../../app/product/containers/ProductFormConnected';

configure({ adapter: new Adapter() });

describe('<ProductFormConnected />', () => {
  let ProductFormWrapper;
  let getProduct;
  let saveProduct;
  let clearCurrentProductInfo;

  const fakeProduct = {
    id: '1',
    imageURL: '#',
    title: 'Mug 1',
    description: 'product 1 description',
    category: { id: '1', name: 'category 1' },
    basePrice: 5,
    currentCurrencyPrice: 5,
    discount: 0,
    producer: 'test producer 1',
    publishDate: 'Tue Aug 20 2019 16:27:02 GMT+0300 (Moscow Standard Time)',
    rate: '5',
  };

  beforeEach(() => {
    getProduct = jest.fn();
    saveProduct = jest.fn();
    clearCurrentProductInfo = jest.fn();

    ProductFormWrapper = shallow(
      <ProductFormConnected
        currentCurrency={{ id: '1', name: 'CURR1', rate: 0.5 }}
        categories={[{ id: '1', name: 'category1' }, { id: '2', name: 'category2' }]}
        getProduct={getProduct}
        saveProduct={saveProduct}
        clearCurrentProductInfo={clearCurrentProductInfo}
      />,
    );
  });

  it('renders spinner with default props and calls getProduct', () => {
    expect(getProduct).toHaveBeenCalled();
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('renders properly with product data provided', () => {
    ProductFormWrapper.setProps({
      productState: {
        product: fakeProduct,
        loadingStatus: { isInProcess: false, hasFailed: false, error: '' },
        savingStatus: { isInProcess: false, hasFailed: false, error: '' },
      },
    });
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('doesnt call getProduct with match.params.id is set to new', () => {
    mount(
      <ProductFormConnected
        currentCurrency={{ id: '1', name: 'CURR1', rate: 0.5 }}
        categories={[{ id: '1', name: 'category1' }, { id: '2', name: 'category2' }]}
        match={{ params: { id: 'new' } }}
        getProduct={getProduct}
        saveProduct={saveProduct}
        clearCurrentProductInfo={clearCurrentProductInfo}
      />,
    );
    expect(getProduct).toHaveBeenCalledTimes(1); // 1 means call getProduct on ProductForm mount defined in beforeEach
  });

  it('renders Spinner when loadingStatus.isInProcess is true', () => {
    ProductFormWrapper.setProps({
      productState: {
        loadingStatus: { isInProcess: true, hasFailed: false, error: '' },
        savingStatus: { isInProcess: false, hasFailed: false, error: '' },
      },
    });
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('renders Spinner when savingStatus.isInProcess is true', () => {
    ProductFormWrapper.setProps({
      productState: {
        loadingStatus: { isInProcess: false, hasFailed: false, error: '' },
        savingStatus: { isInProcess: true, hasFailed: false, error: '' },
      },
    });
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('renders error message when loadingStatus.hasFailed is true', () => {
    ProductFormWrapper.setProps({
      productState: {
        loadingStatus: { isInProcess: false, hasFailed: true, error: 'test product loading error' },
        savingStatus: { isInProcess: false, hasFailed: false, error: '' },
      },
    });
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('renders error message when savingStatus.hasFailed is true', () => {
    ProductFormWrapper.setProps({
      productState: {
        loadingStatus: { isInProcess: false, hasFailed: false, error: '' },
        savingStatus: { isInProcess: false, hasFailed: true, error: 'test product saving error' },
      },
    });
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('calls saveProduct method and passes mapped from view model product', () => {
    ProductFormWrapper.instance().handleSubmit({
      id: '',
      title: 'testTitle',
      imageURL: '###',
      description: 'test description',
      categoryId: '2',
      currentCurrencyPrice: 5,
      discount: 1,
      producer: 'test producer',
      rate: '3',
    });
    expect(saveProduct).toHaveBeenCalledWith({
      product: {
        id: '',
        title: 'testTitle',
        imageURL: '###',
        description: 'test description',
        category: { id: '2', name: 'category2' },
        basePrice: 10,
        discount: 1,
        producer: 'test producer',
        rate: '3',
      },
      redirectUrl: '/catalog',
    });
  });
});
