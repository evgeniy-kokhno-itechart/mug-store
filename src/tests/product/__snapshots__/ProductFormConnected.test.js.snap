// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductFormConnected /> renders Spinner when loadingStatus.isInProcess is true 1`] = `
<Spinner
  spinnerClasses="spinner--medium"
  wrapperClasses="mt-5"
/>
`;

exports[`<ProductFormConnected /> renders Spinner when savingStatus.isInProcess is true 1`] = `
<Spinner
  spinnerClasses="spinner--medium"
  wrapperClasses="mt-5"
/>
`;

exports[`<ProductFormConnected /> renders error message when loadingStatus.hasFailed is true 1`] = `
<ErrorMessage
  customClasses=""
  message="test product loading error"
/>
`;

exports[`<ProductFormConnected /> renders error message when savingStatus.hasFailed is true 1`] = `
<ErrorMessage
  customClasses=""
  message="test product saving error"
/>
`;

exports[`<ProductFormConnected /> renders properly with product data provided 1`] = `
<ProductForm
  categories={
    Array [
      Object {
        "id": "1",
        "name": "category1",
      },
      Object {
        "id": "2",
        "name": "category2",
      },
    ]
  }
  currentCurrencyName="CURR1"
  onSubmit={[Function]}
  product={
    Object {
      "categoryId": "1",
      "currentCurrencyPrice": 5,
      "description": "product 1 description",
      "discount": 0,
      "id": "1",
      "imageURL": "#",
      "producer": "test producer 1",
      "publishDate": "Tue Aug 20 2019 16:27:02 GMT+0300 (Moscow Standard Time)",
      "rate": "5",
      "title": "Mug 1",
    }
  }
/>
`;

exports[`<ProductFormConnected /> renders spinner with default props and calls getProduct 1`] = `
<Spinner
  spinnerClasses="spinner--medium"
  wrapperClasses="mt-5"
/>
`;
